firewood/src/proof.rs

```diff
@@ -1,171 +1,225 @@
 
-    pub fn verify_range_proof<K, V>(
+impl Proof {
+    pub fn verify<K: AsRef<[u8]>, V: AsRef<[u8]>>(
         &self,
-        _root_hash: HashKey,
-        _first_key: K,
-        _last_key: K,
-        keys: Vec<K>,
-        vals: Vec<V>,
-    ) -> Result<bool, ProofError>
-    where
-        K: AsRef<[u8]>,
-        V: AsRef<[u8]>,
-    {
-        if keys.len() != vals.len() {
-            return Err(ProofError::InconsistentProofData);
+        key: K,
+        expected_value: Option<V>,
+        root_hash: &TrieHash,
+    ) -> Result<(), ProofError> {
+        let value_digest = self.value_digest(key, root_hash)?;
+
+        let Some(value_digest) = value_digest else {
+            // This proof proves that `key` maps to None.
+            if expected_value.is_some() {
+                return Err(ProofError::ExpectedValue);
+            }
+            return Ok(());
+        };
+
+        let Some(expected_value) = expected_value else {
+            // We were expecting `key` to map to None.
+            return Err(ProofError::UnexpectedValue);
+        };
+
+        match value_digest {
+            ValueDigest::Value(got_value) => {
+                // This proof proves that `key` maps to `got_value`.
+                if got_value.as_ref() != expected_value.as_ref() {
+                    // `key` maps to an unexpected value.
+                    return Err(ProofError::ValueMismatch);
+                }
+            }
+            ValueDigest::_Hash(got_hash) => {
+                // This proof proves that `key` maps to a value
+                // whose hash is `got_hash`.
+                let value_hash = Sha256::digest(expected_value.as_ref());
+                if got_hash.as_ref() != value_hash.as_slice() {
+                    // `key` maps to an unexpected value.
+                    return Err(ProofError::ValueMismatch);
+                }
+            }
         }
+        Ok(())
+    }
 
-        // Ensure the received batch is monotonic increasing and contains no deletions
-        #[allow(clippy::indexing_slicing)]
-        if !keys.windows(2).all(|w| w[0].as_ref() < w[1].as_ref()) {
-            return Err(ProofError::NonMonotonicIncreaseRange);
+    /// Returns the value digest associated with the given `key` in the trie revision
+    /// with the given `root_hash`. If the key does not exist in the trie, returns `None`.
+    /// Returns an error if the proof is invalid or doesn't prove the key for the
+    /// given revision.
+    fn value_digest<K: AsRef<[u8]>>(
+        &self,
+        key: K,
+        root_hash: &TrieHash,
+    ) -> Result<Option<&ValueDigest<Box<[u8]>>>, ProofError> {
+        let key: Vec<u8> = NibblesIterator::new(key.as_ref()).collect();
+
+        let Some(last_node) = self.0.last() else {
+            return Err(ProofError::Empty);
+        };
+
+        let mut expected_hash = root_hash;
+
+        // TODO danlaine: Is there a better way to do this loop?
+        for i in 0..self.0.len() {
```

Comment:
at a minimum:
```
for (node, i) in self.0.enumerate() {
```

Even better, use a peekable iterator, then you don't need `i` at all:
```
let mut it = self.0.iter().peekable();
while let Some(node) = it.next()  {
   // code for all elements
    if Some(next) = it.peek() {
        // ... code when there is another element in 'next'
    } else {
        // no more elements code
    }
}
```

Reply:
using peek now 👍